source/Attribute.cpp

```diff
@@ -0,0 +1,425 @@
+/* Attribute.cpp
+Copyright (c) 2023 by tibetiroka
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Attribute.h"
+
+#include <algorithm>
+#include <limits>
+#include <unordered_map>
+
+using namespace std;
+
+const string Attribute::effectNames[] = {"shields", "hull", "thrust", "reverse thrust", "turn",
+		"cooling", "active cooling", "cloak", "force", "energy", "fuel", "heat", "discharge", "corrosion", "leak", "burn",
+		"ion", "scramble", "slowing", "disruption", "disabled", "minable", "ramscoop", "piercing", "delay", "depleted delay"};
+
+// Some category names are always the same as the corresponding effect names.
+const string Attribute::categoryNames[] = {"shield generation", "hull repair rate",
+		GetEffectName(THRUST), GetEffectName(REVERSE_THRUST), GetEffectName(TURN), GetEffectName(COOLING),
+		GetEffectName(ACTIVE_COOLING), GetEffectName(CLOAK), "afterburner thrust", "firing", "protection",
+		"resistance", "damage", "capacity"};
+
+bool Attribute::cached = false;
+
+
+
+// Creates a new categorized attribute. Use -1 if there is no category or effect in the definition.
+Attribute::Attribute(AttributeCategory category, AttributeEffect effect, AttributeEffect secondary, bool usePreferred)
+{
+	if(usePreferred)
+	{
+		if(category == -1)
+			category = PASSIVE;
+		if(category == PASSIVE && effect <= CLOAK && effect > HULL)
+			category = static_cast<AttributeCategory>(effect);
+		else if(effect == -1 && category <= CLOAKING)
+			effect = static_cast<AttributeEffect>(category);
+		else if(effect == -1 && category == AFTERBURNING)
+			effect = THRUST;
+		if(category == PASSIVE && effect == PIERCING)
+			category = DAMAGE;
+		if(category == COOL && effect == ENERGY)
+			category = ACTIVE_COOL;
+	}
+	this->category = category;
+	this->effect = effect;
+	this->secondary = secondary;
+}
+
+
+
+// Gets the data format name of the category, as used in the new syntax.
+string Attribute::GetCategoryName(const AttributeCategory category)
+{
+	if(category >= ATTRIBUTE_CATEGORY_COUNT || category < 0)
+		return "";
+	return categoryNames[category];
```

Comment:
```suggestion
	return (category > 0 && category < ATTRIBUTE_CATEGORY_COUNT) ? categoryNames[category] : "";
```
also shouldnt we returning a ref instead of copying it? (and then mb you need to add an empty string thingy to return, idk)

Reply:
Returning a ref is only possible if I make it return a const ref. And in nearly every case, I want the category name to get the full attribute name, in which case, I'm copying it anyway for modifications. I think returning a reference would just make it worse.

Reply:
so many string copies aaaaah

Reply:
This isn't widely used outside of attribute name caching, so I wouldn't worry too much.